File: source/go/barrier_sync.go
package main

import (
	"fmt"
	"sync"
	"sync/atomic"
)

func barrierTask(id int, counter *int32, total int32, doneCh chan<- int, wg *sync.WaitGroup) {
	defer wg.Done()
	atomic.AddInt32(counter, 1)

	// Spin-wait until all tasks have arrived
	for atomic.LoadInt32(counter) < total {
	}

	doneCh <- id
}

func main() {
	var counter int32
	doneCh := make(chan int, 4)
	numTasks := int32(4)

	var wg sync.WaitGroup
	for i := 0; i < int(numTasks); i++ {
		wg.Add(1)
		go barrierTask(i, &counter, numTasks, doneCh, &wg)
	}

	wg.Wait()

	count := 0
	for i := 0; i < int(numTasks); i++ {
		<-doneCh
		count++
	}
	close(doneCh)

	fmt.Printf("all %d tasks passed barrier\n", count)
}

File: source/javascript/barrier_sync.js
// Barrier Synchronization Pattern
// 4 tasks reach a barrier, all proceed after the last arrives.
const { SharedArrayBuffer: SAB } = globalThis;

async function barrierTask(id, counter, total) {
    // Atomically increment
    Atomics.add(counter, 0, 1);
    // Spin-wait until all tasks have arrived
    while (Atomics.load(counter, 0) < total) {
        await new Promise(r => setImmediate(r));
    }
    return id;
}

async function main() {
    const numTasks = 4;
    const sharedBuf = new SharedArrayBuffer(4);
    const counter = new Int32Array(sharedBuf);

    const tasks = [];
    for (let i = 0; i < numTasks; i++) tasks.push(barrierTask(i, counter, numTasks));

    const results = await Promise.all(tasks);
    const count = results.length;
    console.log(`all ${count} tasks passed barrier`);
}

main();
